File: instagram-backend/src/main/java/com/hoanghuy04/instagrambackend/service/comment/CommentService.java
package com.hoanghuy04.instagrambackend.service.comment;

import com.hoanghuy04.instagrambackend.dto.request.CreateCommentRequest;
import com.hoanghuy04.instagrambackend.dto.response.CommentResponse;
import com.hoanghuy04.instagrambackend.dto.response.PageResponse;
import org.springframework.data.domain.Pageable;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.List;

/**
 * Service interface for comment operations.
 * Handles comment creation, updates, and queries.
 * 
 * @author Instagram Backend Team
 * @version 1.0.0
 */
@Service
public interface CommentService {
    
    /**
     * Create a new comment on a post.
     *
     * @param request the comment creation request
     * @return CommentResponse
     */
    @Transactional
    CommentResponse createComment(CreateCommentRequest request);

    /**
     * Get comment by ID.
     *
     * @param commentId the comment ID
     * @return CommentResponse
     */
    @Transactional(readOnly = true)
    CommentResponse getComment(String commentId);

    /**
     * Get comments for a post.
     *
     * @param postId the post ID
     * @param pageable pagination information
     * @return PageResponse of CommentResponse
     */

    /**
     * Update a comment.
     *
     * @param commentId the comment ID
     * @param text the new comment text
     * @return updated CommentResponse
     */
    @Transactional
    CommentResponse updateComment(String commentId, String text);

    /**
     * Delete a comment.
     *
     * @param commentId the comment ID
     */
    @Transactional
    void deleteComment(String commentId);

    /**
     * Reply to a comment.
     *
     * @param commentId the comment ID to reply to
     * @param request the comment request
     * @return CommentResponse
     */
    @Transactional
    CommentResponse replyToComment(String commentId, CreateCommentRequest request);

    /**
     * Like a comment.
     *
     * @param commentId the comment ID
     */
    @Transactional
    boolean toggleLikeComment(String commentId);

    @Transactional(readOnly = true)
    PageResponse<CommentResponse> getPostComments(String postId, Pageable pageable);

    @Transactional
    List<CommentResponse> getCommentReplies(String commentId);

}

File: instagram-frontend-mobile/services/comment.service.ts
import { commentAPI } from './api';
import { Comment, CreateCommentRequest, PaginatedResponse } from '@types';

/**
 * Comment service with enhanced mention handling
 */
class CommentService {
  /**
   * Get comments for a post (root comments only)
   */
  async getComments(
    postId: string,
    page = 0,
    limit = 20
  ): Promise<PaginatedResponse<Comment>> {
    return commentAPI.getComments(postId, page, limit);
  }

  /**
   * Get replies for a specific comment
   */
  async getReplies(commentId: string): Promise<Comment[]> {
    return commentAPI.getReplies(commentId);
  }

  /**
   * Create a root comment
   */
  async createComment(data: CreateCommentRequest): Promise<Comment> {
    return commentAPI.createComment(data);
  }

  /**
   * Reply to a comment with automatic mention handling
   */
  async replyToComment(
    parentCommentId: string, 
    data: CreateCommentRequest
  ): Promise<Comment> {
    return commentAPI.replyToComment(parentCommentId, data);
  }

  /**
   * Update comment text
   */
  async updateComment(commentId: string, text: string): Promise<Comment> {
    return commentAPI.updateComment(commentId, text);
  }

  /**
   * Delete a comment
   */
  async deleteComment(commentId: string): Promise<void> {
    return commentAPI.deleteComment(commentId);
  }

  /**
   * Toggle like on comment
   */
  async toggleLikeComment(commentId: string): Promise<boolean> {
    return commentAPI.toggleLikeComment(commentId);
  }
}

export const commentService = new CommentService();
